src/utils.c

```diff
@@ -108,6 +115,80 @@ R_len_t compact_rownames_length(SEXP x) {
   return abs(INTEGER(x)[1]);
 }
 
+static void init_compact_rownames(SEXP x, R_len_t n);
+static SEXP new_compact_rownames(R_len_t n);
+
+void init_data_frame(SEXP x, R_len_t n) {
+  Rf_setAttrib(x, R_ClassSymbol, classes_data_frame);
+  init_compact_rownames(x, n);
+}
+void init_tibble(SEXP x, R_len_t n) {
+  Rf_setAttrib(x, R_ClassSymbol, classes_tibble);
+  init_compact_rownames(x, n);
+}
+
+static void init_compact_rownames(SEXP x, R_len_t n) {
+  SEXP rn = PROTECT(new_compact_rownames(n));
+  Rf_setAttrib(x, R_RowNamesSymbol, rn);
+  UNPROTECT(1);
+}
+static SEXP new_compact_rownames(R_len_t n) {
+  if (n <= 0) {
+    return vctrs_shared_empty_int;
+  }
+
+  SEXP out = Rf_allocVector(INTSXP, 2);
+  int* out_data = INTEGER(out);
+  out_data[0] = NA_INTEGER;
+  out_data[1] = -n;
+  return out;
+}
+
+
+static bool is_tibble_class(SEXP class);
+static bool is_data_frame_class(SEXP class);
+
+bool is_tibble(SEXP x) {
+  SEXP class = PROTECT(Rf_getAttrib(x, R_ClassSymbol));
+  bool out = is_tibble_class(class);
+  UNPROTECT(1);
+  return out;
+}
+bool is_native_df(SEXP x) {
+  SEXP class = PROTECT(Rf_getAttrib(x, R_ClassSymbol));
+  bool out = is_data_frame_class(class) || is_tibble_class(class);
+  UNPROTECT(1);
+  return out;
+}
+
+static bool is_tibble_class(SEXP class) {
+  if (Rf_length(class) != 3) {
+    return false;
+  }
+
+  SEXP* class_ptr = STRING_PTR(class);
+
+  if (*class_ptr != strings_tbl_df) {
+    return false;
+  }
+  ++class_ptr;
+  if (*class_ptr != strings_tbl) {
+    return false;
+  }
+  ++class_ptr;
+  if (*class_ptr != strings_data_frame) {
+    return false;
+  }
+
+  return true;
```

Comment:
Why doesn't this need to fall back to `Rf_inherits()`?

Reply:
This is used for optimisating objects that are actually tibbles. Objects inheriting from it should call `vec_restore()`. The dispatch path for tibbles is still not entirely clear to me though.

I'll add a `dispatch` parameter so the semantics at the call site are clearer.

Reply:
This is now `is_bare_tibble()`.

Reply:
Actually I was confused by the flow, because you're not first comparing `class` to a cached character vector. Or did we decide that optimisation was not worth it?

Reply:
This would require tibble to use a constructor exported in vctrs. Comparing the length and elements of the class should be fast as well.

Reply:
Also this method works with unserialised objects.

Reply:
Right, but I think that's irrelevant compared to the cost of having a subtly different way of checking for different classes. It's better to standardise on a single approach and use it everywhere.